### PR TITLE
perf(0.39.0): bundle — BQ query rewrite + session pool + chunked download + HTTP/2

### PR DESCRIPTION
**Pull request by ZdenekSrotyr**
## Summary

Bundle of perf fixes targeting two user-visible pain points measured in production diagnostics today:

- **`agnes query --remote` 50–100× slower** than direct `bigquery_query()` against the same data. Master view ATTACH-catalog mode didn't push WHERE/SELECT/LIMIT into BQ — opened Storage Read API session over the entire upstream table, often failed with `Response too large to return`.
- **`agnes pull` for a 5.1 GB single materialized parquet ran 0.29 MB/s** single-stream over corp VPN with per-TCP-connection rate-limiting. 4 parallel range requests sustain 1.65 MB/s aggregate (5.7× linear scale).

### Server-side (`app/api/query.py`, `connectors/bigquery/access.py`)

- **`bigquery_query()` rewrite on execution path** — extracts the existing `_rewrite_user_sql_for_bq_dry_run` core into shared helper, applies it on `analytics.execute(...)` so user SQL referencing remote BQ tables is wrapped in `bigquery_query('<project>', '<inner-sql>')` before execute. BQ planner gets the full query, applies partition pruning + projection pushdown server-side. Conservative fall-through: cross-source JOINs (BQ × Keboola/Jira local), queries already containing `bigquery_query(`, and unconfigured BQ project all keep the original ATTACH path.
- **Outer LIMIT wrap** — `bigquery_query()` materialises BQ result into DuckDB before fetchmany sees it; outer `LIMIT N+1` pushes the cap into the BQ job itself so a `SELECT *` against a billion-row table doesn't OOM the worker.
- **Dollar-quoted inner SQL** — `$bqq_inner$ … $bqq_inner$` form takes the inner SQL verbatim, no escape sequences whatsoever (vs naive `replace("'", "''")` which missed DuckDB backslash escapes).
- **Parse-error fallback** — if the rewritten path fails with a BQ-side parse/validation error (e.g. user SQL had DuckDB-only `::INT` cast that survived identifier rewrite but BQ refuses), retry the user's original SQL via the legacy ATTACH path so the request still succeeds.
- **DuckDB BQ extension session pool** — `BqAccess.duckdb_session()` acquires pre-warmed connections from a bounded process-local pool instead of running `INSTALL bigquery; LOAD bigquery; CREATE SECRET; ATTACH …` on every request. Saves ~0.5 s per BQ query. Pool size configurable via `data_source.bigquery.session_pool_size` (default 4; sentinel `0` disables pooling).
- **`responseTooLarge` error mapping** — was previously misclassified as generic `bq_bad_request` (implying SQL syntax error), now has a dedicated `bq_response_too_large` kind with actionable hint pointing at WHERE / aggregation / materialized-table remediations.

### CLI-side (`cli/client.py`, `cli/lib/pull.py`)

- **Range-chunked download** — when server advertises `accept-ranges: bytes` and a parquet exceeds `AGNES_PULL_CHUNK_THRESHOLD_BYTES` (default 50 MB), CLI splits into N parallel HTTP Range requests (`AGNES_PULL_CHUNK_PARALLELISM`, default 4, capped 1..16) and assembles atomically. Targets per-flow-shaped networks where 1 stream is throttled but N parallel streams over the same connection scale linearly.
- **Persistent HTTP/2 client** — process-wide pooled `httpx.Client` so N parquet downloads share a single TLS handshake; HTTP/2 multiplexing (when `h2` package installed) lets all chunk Range requests share one TCP connection. Graceful fallback to HTTP/1.1 when `h2` missing.
- **Textual progress fallback for non-TTY** — when stderr is not a terminal (Claude Code SessionStart hook, CI runner, Docker log capture), `agnes pull` emits plain-text progress per file at most every 10% or 30 s. Replaces previous Rich-bar-on-pipe behavior that either suppressed output entirely or leaked ANSI sequences.

### Release

- `pyproject.toml` 0.38.3 → 0.39.0
- CHANGELOG `[Unreleased]` → `[0.39.0] — 2026-05-06`
- New `[Unreleased]` empty section appended

## Expected impact

- `agnes query --remote` on a typical aggregate against a 100M+ row remote BQ view: **70–150 s → ~1.5 s**
- `agnes pull` of single 5 GB parquet over per-flow-throttled VPN: **5 hours → ~50 min** (1.65 MB/s aggregate vs 0.29 MB/s single)
- BQ extension session reuse saves ~0.5 s per `agnes query --remote`, `/api/v2/scan`, `/api/v2/sample`, `/api/v2/schema`

## Test plan

- [x] 71 perf-related tests pass (12 rewriter + 5 pool + 4 translate-error + 9 BQ-existing + 7 chunked + 3 progress + 3 shared-client + edge cases)
- [x] Code review by `superpowers:code-reviewer` — 2 MUST blockers + 2 SHOULDs identified, all addressed in fix commit `8e56d45c`
- [ ] Smoke-test against production after merge: `agnes query --remote --sql "SELECT COUNT(*) FROM s1_session_landings WHERE event_date = ..."` should return in ~1.5 s
- [ ] Smoke-test `agnes pull` of `order_economics` (5.1 GB) — should complete in 30–60 min on per-flow-throttled link, 30–60 s on unthrottled

## Risk

- **Rewriter behavior change** is the primary regression surface. Queries with DuckDB-only syntax (`::INT`, `STRPTIME`, COALESCE arity differences) that survived identifier rewrite would fail under `bigquery_query()`. Fallback in handler retries via legacy path, so user requests still succeed — just slower than the rewrite. Heuristic catches "Syntax error", "Unrecognized name", "Function not found", "No matching signature", "Invalid cast" patterns.
- **HTTP/2 fallback** — if `h2` package install fails on a fresh VM, CLI falls back to HTTP/1.1 with connection pooling. Verified via monkeypatch test; real-world fresh-install verification still TODO post-merge.
- **Pool docstring caveat** — `session_pool_size` sizes the *retain-on-release* idle cache, not concurrent in-flight sessions. High-concurrency burst (~100+ simultaneous BQ queries) would still create N independent DuckDB sessions per request. Production single-worker uvicorn naturally serializes requests, so this is not currently a memory risk; documented for future readers.

## Files modified

- `app/api/query.py` (+260 LOC: rewriter helper extraction + handler integration + fallback)
- `connectors/bigquery/access.py` (+280 LOC: session pool + translate_bq_error fix)
- `cli/client.py` (+400 LOC: chunked download + shared HTTP/2 client)
- `cli/lib/pull.py` (+170 LOC: progress orchestration + non-TTY fallback)
- `pyproject.toml` (`h2>=4.1.0` dep + version bump)
- `config/instance.yaml.example` (`session_pool_size` config knob)
- 5 new test files (~1,775 LOC)
- `CHANGELOG.md` (release cut + duplicate cleanup)

---

<a href="https://app.devin.ai/review/keboola/agnes-the-ai-analyst/pull/199" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by ZdenekSrotyr** [comment]
@devin-ai-integration please re-review the latest pushes — 4 fix commits since your initial review on `8e56d45c`:

- `e5645fd2` devil's-advocate R1 fixes (5)
- `aee585fa` devil's-advocate R2 fixes (3)
- `77d88014` devil's-advocate R3 fix (1)
- `81d065b1` Devin Review #1 fix (`bigquery_query()` first-arg uses billing project)

Devin Review #2 (overly broad try/except in `stream_download`) was already addressed by R2 commit `aee585fa` before your initial review fully landed.

File: app/api/query.py [hunk 1]
```diff
@@ -509,6 +597,180 @@ def _bq_path_repl(m: re.Match) -> str:
...
+                "rewrite_skip_cross_source: user SQL references both "
+                "BQ-remote and local-mode tables; falling back to "
+                "ATTACH-catalog path",
+            )
+            return user_sql, False
+
+    # Skip 4: BQ project not configured.
+    try:
+        bq = get_bq_access()
+        project = bq.projects.data
```
[the remark this answers is not in the document]
Comment (ZdenekSrotyr) [reply app/api/query.py]: Fixed in commit `81d065b1`. `bigquery_query()` first arg now uses `bq.projects.billing` (matching the convention in `v2_scan`/`v2_sample`/`v2_schema`/`extractor`); inner backtick paths still use `bq.projects.data`. Test pinned at `tests/test_query_remote_rewrite.py:test_rewriter_uses_billing_project_for_bigquery_query_first_arg` (cross-project case: `data-prj` vs `billing-prj`).

File: cli/client.py [hunk 2]
```diff
@@ -197,57 +292,292 @@ def _is_transient(exc: Exception) -> bool:
...
+    # handshake amortized across N stream_download calls within the same
+    # process, and HTTP/2 stream multiplexing across the chunk Range
+    # requests within a single download. Falls back to a fresh per-call
+    # client if shared-client construction fails for any reason.
+    try:
+        client = _get_shared_client()
+        return _stream_download_via(client, path, target_path, progress_callback)
+    except Exception:
+        with get_client(timeout=300.0) as client:
+            return _stream_download_via(client, path, target_path, progress_callback)
```
[the remark this answers is not in the document]
Comment (ZdenekSrotyr) [reply cli/client.py:658]: Already addressed in commit `aee585fa` (devil's-advocate R2 fix #1) before your initial review fully landed. The try/except is now scoped to **only** the `_get_shared_client()` construction; the actual download is no longer retried under the fallback except. See `cli/client.py:594-599`.
